pkg/backend/httpstate/client/api.go

```diff
@@ -172,21 +209,19 @@ func pulumiAPICall(ctx context.Context, d diag.Sink, cloudAPI, method, path stri
 
 	// For 4xx and 5xx failures, attempt to provide better diagnostics about what may have gone wrong.
 	if resp.StatusCode >= 400 && resp.StatusCode <= 599 {
-		defer contract.IgnoreClose(resp.Body)
-
-		// Provide a better error if using an authenticated call without having logged in first.
-		if resp.StatusCode == 401 && tok.Kind() == accessTokenKindAPIToken && tok.String() == "" {
-			return "", nil, errors.New("this command requires logging in; try running 'pulumi login' first")
-		}
-
 		// 4xx and 5xx responses should be of type ErrorResponse. See if we can unmarshal as that
 		// type, and if not just return the raw response text.
-		respBody, err := ioutil.ReadAll(resp.Body)
+		respBody, err := readBody(resp)
 		if err != nil {
 			return "", nil, errors.Wrapf(
 				err, "API call failed (%s), could not read response", resp.Status)
 		}
 
+		// Provide a better error if using an authenticated call without having logged in first.
+		if resp.StatusCode == 401 && tok.Kind() == accessTokenKindAPIToken && tok.String() == "" {
+			return "", nil, errors.New("this command requires logging in; try running 'pulumi login' first")
```

Comment:
Was this just to streamline an error case you hit while testing this locally? IIRC we already have some hooks somewhere to provide a similar message for login/auth errors like this. So I'm reluctant to add another way to do the same thing, at least not without understanding why the redundancy is needed. (Or if I'm totally mistaken and we have no such error.)

Reply:
This code is not new, it's moved from above the call to `readBody` to be below it. The code was previously inspecting the body before decoding it, which would be problematic if the service's response was gzipped. Here we are reading the body using `readBody` so that we can be graceful regardless of the encoding of the response.